tests/test_mpls.py

```diff
@@ -0,0 +1,591 @@
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"ip route 2.2.2.0/24 10.0.0.1 Ethernet0 label 201\"")
+
+        # check application database
+        self.pdb.wait_for_entry("ROUTE_TABLE", prefix)
+        time.sleep(1)
+
+        # check ASIC route database
+        self.check_route_entries(True, [prefix])
+        self.check_route_nexthop(prefix, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_PUSH", "1:201")
+
+        # remove route entry
+        if self.mpls_appdb_mode():
+            self.remove_route_entry(prefix)
+        else:
+            # dvs.runcmd("ip route del 2.2.2.0/24 encap mpls 201 via inet 10.0.0.1 dev Ethernet0")
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"no ip route 2.2.2.0/24 10.0.0.1 Ethernet0 label 201\"")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("ROUTE_TABLE", prefix)
+        time.sleep(1)
+
+        # check ASIC route database
+        self.check_route_entries(False, [prefix])
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_PUSH", "1:201")
+
+        self.teardown_mpls(dvs)
+
+    def test_RouteAddRemoveMplsRouteSwap(self, dvs, testlog):
+        self.setup_mpls(dvs, True)
+
+        # add route entry
+        label = "200"
+        if self.mpls_appdb_mode():
+            fieldValues = {"nexthop": "10.0.0.1", "ifname": "Ethernet0", "mpls_nh": "swap+201", "mpls_pop": "1"}
+            self.create_inseg_entry(label, fieldValues)
+        else:
+            # dvs.runcmd("ip -f mpls route add 200 as 201 via inet 10.0.0.1 dev Ethernet0")
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"mpls lsp 200 10.0.0.1 201\"")
+
+        # check application database
+        self.pdb.wait_for_entry("LABEL_ROUTE_TABLE", label)
+        time.sleep(1)
+
+        # check ASIC inseg database
+        self.check_inseg_entries(True, [label])
+        self.check_inseg_nexthop(label, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_SWAP", "1:201")
+
+        # remove route entry
+        if self.mpls_appdb_mode():
+            self.remove_inseg_entry(label)
+        else:
+            # dvs.runcmd("ip -f mpls route del 200 as 201 via inet 10.0.0.1 dev Ethernet0")
+            dvs.runcmd("vtysh -c \"configure terminal\" -c \"no mpls lsp 200 10.0.0.1 201\"")
+
+        # check application database
+        self.pdb.wait_for_deleted_entry("LABEL_ROUTE_TABLE", label)
+        time.sleep(1)
+
+        # check ASIC inseg database
+        self.check_inseg_entries(False, [label])
+        self.check_nexthop(False, "SAI_NEXT_HOP_TYPE_MPLS", "10.0.0.1", "SAI_OUTSEG_TYPE_SWAP", "1:201")
+
+        self.teardown_mpls(dvs)
+
+    def test_RouteAddRemoveMplsRoutePop(self, dvs, testlog):
+        self.setup_mpls(dvs, True)
+
+        # add route entry
+        label = "200"
+        if self.mpls_appdb_mode():
+            fieldValues = {"nexthop": "10.0.0.1", "ifname": "Ethernet0", "mpls_nh": "swap+", "mpls_pop": "1"}
```

Comment:
is "swap+" required here?

Reply:
@smahesh Yes, you're the one who stated that POP was programmed in SAI as swap action with empty labelstack.

Reply:
Oh no. POP is always implicit unless there's a PUSH action. If there's no label in the nexthop there's no need to add any action. May be your code also needs an update, can you check.

Reply:
@smaheshm Programming is updated per latest discussion.